### ISSUE TITLE
save_weights and load_weights do not work as expected when used to restore original initialized weights

### ISSUE THREAD

**Issue by amaiya** [tensorflow/tensorflow#41116]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): YES
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.6.9

**Describe the current behavior**

The `save_weights` and `load_weights` functions do not work as expected when used to save and restore the original, initialized weights of a model. For instance, in the following example, we create a model, save the original weights, corrupt the weights, and restore the original, uncorrupted weights.  The training results in a very high loss when the `corrupt_weights_but_restore` function is invoked, but a low loss when the `corrupt_weights_but_restore` is commented out.

```python
import tempfile
from tensorflow.keras import backend as K
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.datasets import boston_housing
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()
def corrupt_weights_but_restore(model, X, Y):

   # save original uncorrupted weights
    new_file, weightfile = tempfile.mkstemp()
    model.save_weights(weightfile)       

   # corrupt/damage weights with very high learning rate
    K.set_value(model.optimizer.lr, 100) 
    model.fit(X, Y, epochs=5)

   # load original uncorrupted weights
    model.load_weights(weightfile) 
    return model

# build model
model = Sequential()
model.add(Dense(1, input_shape=(x_train.shape[1],), activation='linear'))
model.compile(optimizer='adam', loss='mse', metrics=['mse', 'mae'])

# uncommenting the line below produces high loss despite restoring original weights
model = corrupt_weights_but_restore(model, x_train, y_train)

# train
K.set_value(model.optimizer.lr, 0.05)
model.fit(x_train, y_train,
          batch_size=32,
          epochs=100)
~                       
```

**Describe the expected behavior**
Loss should be the same (and be low) regardless of whether or not `corrupt_weights_but_restore` is run, since the original, uncorrupted weights are restored in `corrupt_weights_but_restore`.

**Comment by ravikyram** [comment]
@amaiya 

I have tried in colab with TF version 2.2.Please, find the gist here. You are also seeing the same behavior?.Thanks!

**Comment by amaiya** [comment]
@ravikyram Sorry - where exactly is the gist?  I don't see a link i your post.

In TensorFlow 2.2.0, when the `corrupt_weights_but_restore` method is commented out, the training loss is around 30.  When it is uncommented and invoked, the loss is several hundred.  I've reproduced this across several different models and datasets.

**Comment by ravikyram** [comment]
@amaiya 

Sorry, my bad. Forgot to attach [gist](https://colab.research.google.com/gist/ravikyram/5ec20121376d550db67784ae6aeca738/untitled84.ipynb).You are also seeing the same behavior?Thanks!

**Comment by amaiya** [comment]
Thanks - no worries.   Yes that's the behavior I'm seeing.  Now, if you do either the following, the problem disappears and the loss is much lower (between 30 and 40 versus the loss of 702 shown in your gist):
- comment out the line that calls `corrupt_weights_but_restore`
- replace `save_weights` and `load_weights` with `model.save` and `load_model`, respectively

This suggests that `save_weights` and `load_weights` are not doing the right thing, which is problematic.

See [this Google Colab](https://colab.research.google.com/drive/19K5PxnxOxV00AdBQWLwB7Rc3-KWu_R0b?usp=sharing) that shows this.

**Comment by ravikyram** [comment]
I have tried in colab with nightly version(`2.4.0-dev20200706`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/85dd12dbec8001a129e888d7b1392197/untitled87.ipynb).Thanks!

**Comment by k-w-w** [comment]
Just a note -- h5 doesn't restore optimizer weights, so if the loss after training is extremely high, that could mean that the adam optimizer weights are corrupted from calling `model.fit()` on the high learning rate. I would suggest using the TF format, or recompiling the model after loading the weights.

**Comment by amaiya** [comment]
@k-w-w :  Thank you - it looks like you're exactly right about what's happening.  Thanks for the suggestions.  I am closing this issue
